fe/fe-core/src/main/java/com/starrocks/clone/ColocateTableBalancer.java

```diff
@@ -357,38 +357,44 @@ private boolean relocateAndBalance(GroupId groupId, Set<Long> unavailableBeIds,
                     break;
                 }
             }
-            // sort backends with replica num in desc order
+
+            // Sort backends with replica num in desc order, the list contains only all the *available* backends.
             List<Map.Entry<Long, Long>> backendWithReplicaNum =
                     getSortedBackendReplicaNumPairs(availableBeIds, unavailableBeIds, statistic,
                             flatBackendsPerBucketSeq);
-            if (srcBeSeqIndexes == null || srcBeSeqIndexes.size() <= 0) {
-                // there is only one available backend and no unavailable bucketId to relocate, end the outer loop
-                if (backendWithReplicaNum.size() <= 1) {
-                    break;
-                }
-
-                // there is no unavailable bucketId to relocate, choose max bucketId num be as src be
-                srcBeId = backendWithReplicaNum.get(0).getKey();
-                srcBeSeqIndexes = getBeSeqIndexes(flatBackendsPerBucketSeq, srcBeId);
-            } else if (backendWithReplicaNum.size() <= 0) {
-                // there is unavailable bucketId to relocate, but no available backend, end the outer loop
+            if (backendWithReplicaNum.size() <= 1) {
+                // There is not enough replicas for us to do relocation or balance, because in this case we
+                // can not choose a valid backend to migrate replica to, end the outer loop.
                 break;
             }
 
             int leftBound;
-            if (hasUnavailableBe) {
-                leftBound = -1;
-            } else {
+            if (!hasUnavailableBe || replicationNum == 1) {
```

Comment:
Unavailable BE means: 1. stopped, 2. decommissioned.  For the decommissioned case, we should migrate the bucketId on that BE, or the decommission will be blocked.

Reply:
Yes, i should consider this situation, let me fix that.